pkg/workloads/cortex/lib/model/model.py

```diff
@@ -0,0 +1,596 @@
+
+        self._models = {}
+        self._timestamps = {}
+        self._locks = {}
+
+        self._create_lock = td.RLock()
+        self._global_lock = ReadWriteLock()
+
+    def set_callback(self, ctype: str, callback: Callable) -> None:
+        """
+        Sets a callback.
+
+        Args:
+            ctype: "download", "load" or "remove" callback type - see the constructor to mark each one.
+            callback: The actual callback.
+        """
+        if ctype == "download":
+            self._download_callback = callback
+        if ctype == "load":
+            self._load_callback = callback
+        if ctype == "remove":
+            self._remove_callback = callback
+
+    #################
+
+    def global_acquire(self, mode: str, timeout: Optional[float] = None) -> None:
+        """
+        Acquire shared/exclusive (R/W) access over all models.
+
+        Use "w" when wanting to acquire exclusive access for the GC (method garbage_collect), or "r" when wanting to grant shared access for any other method to be called (i.e. get_model_ids).
+
+        Args:
+            mode: "r" for read lock, "w" for write lock.
+            timeout: How many seconds to wait to acquire the lock.
+        """
+        self._global_lock.acquire(mode)
+
+    def global_release(self, mode: str) -> None:
+        """
+        Release shared/exclusive (R/W) access over all models.
+
+        Args:
+            mode: "r" for read lock, "w" for write lock.
+        """
+        self._global_lock.release(mode)
+
+    def model_acquire(self, mode: str, model_name: str, model_version: str) -> None:
+        """
+        Acquire shared/exclusive (R/W) access for a specific model.
+
+        Args:
+            mode: "r" for read lock, "w" for write lock.
+            model_name: The name of the model.
+            model_version: The version of the model.
+
+        When mode is "r", only the following methods can be called:
+        * has_model
+        * get_model
+
+        When mode is "w", the methods available for "r" can be called plus the following ones:
+        * load_model
+        * download_model
+        * remove_model
+        """
+        model_id = f"{model_name}-{model_version}"
+
+        if not model_id in self._locks:
+            lock = ReadWriteLock()
+            lock_id = id(lock)
+            self._create_lock.acquire()
+            if model_id not in self._locks:
+                self._locks[model_id] = lock
+            self._create_lock.release()
+
+        self._locks[model_id].acquire(mode)
+
+    def model_release(self, mode: str, model_name: str, model_version: str) -> None:
+        """
+        Release shared/exclusive (R/W) access for a specific model.
+
+        Args:
+            mode: "r" for read lock, "w" for write lock.
+            model_name: The name of the model.
+            model_version: The version of the model.
+        """
+        model_id = f"{model_name}-{model_version}"
```

Comment:
When will the locks in self._locks get removed?

Reply:
Never.